cmd/kindnetd/dnscache.go

```diff
@@ -0,0 +1,741 @@
+		klog.V(7).Infof("DNS A request for %s", q.Name.String())
+		addrs, err := d.lookupIP(context.Background(), "ip4", q.Name.String())
+		if err != nil {
+			klog.V(7).Infof("DNS A request for %s error: %v", q.Name.String(), err)
+			return dnsErrorMessage(id, dnsmessage.RCodeServerFailure, q), false
+		}
+		if len(addrs) == 0 {
+			return dnsErrorMessage(id, dnsmessage.RCodeNameError, q), false
+		}
+		klog.V(7).Infof("DNS A request for %s ips: %v", q.Name.String(), addrs)
+		for _, ip := range addrs {
+			a := ip.To4()
+			if a == nil {
+				continue
+			}
+			err = answer.AResource(
+				dnsmessage.ResourceHeader{
+					Name:  q.Name,
+					Class: q.Class,
+					TTL:   ttl,
+				},
+				dnsmessage.AResource{
+					A: [4]byte{a[0], a[1], a[2], a[3]},
+				},
+			)
+			if err != nil {
+				return dnsErrorMessage(id, dnsmessage.RCodeServerFailure, q), false
+			}
+		}
+	case dnsmessage.TypeAAAA:
+		klog.V(7).Infof("DNS AAAA request for %s", q.Name.String())
+		addrs, err := d.lookupIP(context.Background(), "ip6", q.Name.String())
+		if err != nil {
+			klog.V(7).Infof("DNS AAAA request for %s error: %v", q.Name.String(), err)
+			return dnsErrorMessage(id, dnsmessage.RCodeServerFailure, q), false
+		}
+		if len(addrs) == 0 {
+			return dnsErrorMessage(id, dnsmessage.RCodeNameError, q), false
+		}
+		klog.V(7).Infof("DNS AAAA request for %s ips: %v", q.Name.String(), addrs)
+		for _, ip := range addrs {
+			if ip.To16() == nil || ip.To4() != nil {
+				continue
+			}
+			var aaaa [16]byte
+			copy(aaaa[:], ip.To16())
+			err = answer.AAAAResource(
+				dnsmessage.ResourceHeader{
+					Name:  q.Name,
+					Class: q.Class,
+					TTL:   ttl,
+				},
+				dnsmessage.AAAAResource{
+					AAAA: aaaa,
+				},
+			)
+			if err != nil {
+				return dnsErrorMessage(id, dnsmessage.RCodeServerFailure, q), false
+			}
+		}
+	case dnsmessage.TypePTR:
+		klog.V(7).Infof("DNS PTR request for %s", q.Name.String())
+		return nil, true
+	case dnsmessage.TypeSRV:
+		return nil, true
+	case dnsmessage.TypeNS:
+		return nil, true
+	case dnsmessage.TypeCNAME:
+		return nil, true
+	case dnsmessage.TypeSOA:
+		return nil, true
+	case dnsmessage.TypeMX:
+		return nil, true
+	case dnsmessage.TypeTXT:
+		return nil, true
+	default:
+		return nil, true
+	}
+	buf, err = answer.Finish()
+	if err != nil {
+		return dnsErrorMessage(id, dnsmessage.RCodeServerFailure, q), false
+	}
+	return buf, false
+}
+
+func (d *DNSCacheAgent) lookupIP(ctx context.Context, network, host string) ([]net.IP, error) {
+	ips, ok := d.cache.get(network, host, expireTimeout)
+	if ok {
+		klog.V(4).Infof("Cached entries for %s %s : %v", network, host, ips)
+		return ips, nil
+	}
+	ips, err := d.resolver.LookupIP(ctx, network, host)
+	if err != nil {
+		// cache empty answers
+		if e, ok := err.(*net.DNSError); !ok || !e.IsNotFound {
+			return nil, err
+		}
+	}
+	d.cache.add(network, host, ips)
+	klog.V(4).Infof("Caching new entries for %s %s : %v", network, host, ips)
+	return ips, nil
+}
```

Comment:
This is the important thing to solve:
- cache agent has the pod parameters ndots and searches (it is deployed as a hostnetwork pod with dnsPoliyc: ClusterFirstWithHostNet)
- when pod ask for `kubernetes.default` the resolver on the pod will append all `searches` until the number of `ndots`

```
I0913 09:47:13.869272       1 dnscache.go:249] UDP connection from 100.96.1.14:59154
I0913 09:47:13.869338       1 dnscache.go:283] dialing from 100.64.0.10:53 to 100.96.1.14:59154
I0913 09:47:13.869435       1 dnscache.go:444] starting parsing packet
I0913 09:47:13.869469       1 dnscache.go:559] DNS A request for kubernetes.default.default.svc.cluster.local.
I0913 09:47:13.869484       1 dnscache.go:658] Cached entries for ip4 kubernetes.default.default.svc.cluster.local. : []
I0913 09:47:13.869498       1 dnscache.go:496] answer correct
I0913 09:47:13.870417       1 dnscache.go:249] UDP connection from 100.96.1.14:52573
I0913 09:47:13.870465       1 dnscache.go:283] dialing from 100.64.0.10:53 to 100.96.1.14:52573
I0913 09:47:13.870550       1 dnscache.go:444] starting parsing packet
I0913 09:47:13.870578       1 dnscache.go:559] DNS A request for kubernetes.default.svc.cluster.local.
I0913 09:47:13.871531       1 dnscache.go:669] Caching new entries for ip4 kubernetes.default.svc.cluster.local. : [100.64.0.1]
I0913 09:47:13.871627       1 dnscache.go:568] DNS A request for kubernetes.default.svc.cluster.local. ips: [100.64.0.1]
I0913 09:47:13.871640       1 dnscache.go:496] answer correct
I0913 09:47:13.874943       1 dnscache.go:249] UDP connection from 100.96.1.14:55191
I0913 09:47:13.874996       1 dnscache.go:283] dialing from 100.64.0.10:53 to 100.96.1.14:55191
I0913 09:47:13.875084       1 dnscache.go:444] starting parsing packet
I0913 09:47:13.875150       1 dnscache.go:589] DNS AAAA request for kubernetes.default.svc.cluster.local.
I0913 09:47:13.876114       1 dnscache.go:669] Caching new entries for ip6 kubernetes.default.svc.cluster.local. : []
I0913 09:47:13.876212       1 dnscache.go:496] answer correct
```

This means that the cache CAN know the original question of the pod and can deduplicate all these queries to answer on the first query ... this will be a big improvement since it will allow to remove all the amplification added by `searches` @thockin